File: backend/src/main/java/vn/student/polyshoes/service/CategoryService.java
package vn.student.polyshoes.service;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import vn.student.polyshoes.dto.CategoryDto;
import vn.student.polyshoes.model.Category;
import vn.student.polyshoes.repository.CategoryRepository;
import vn.student.polyshoes.repository.SubCategoryRepository;
import vn.student.polyshoes.response.CategoryResponse;

import java.util.Date;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Service
public class CategoryService {

    @Autowired
    private CategoryRepository categoryRepository;

    @Autowired
    private SubCategoryRepository subCategoryRepository;

    public List<CategoryResponse> getAllCategories() {
        return categoryRepository.findAll().stream()
                .map(this::mapToCategoryResponse)
                .collect(Collectors.toList());
    }

    public List<CategoryResponse> getActiveCategories() {
        return categoryRepository.findByIsActiveTrue().stream()
                .map(this::mapToCategoryResponse)
                .collect(Collectors.toList());
    }

    public CategoryResponse getCategoryById(Integer categoryId) {
        return categoryRepository.findById(categoryId)
                .map(this::mapToCategoryResponse)
                .orElse(null);
    }

    public boolean isCategoryNameExists(String categoryName) {
        return categoryRepository.existsByCategoryName(categoryName);
    }

    public boolean isCategoryNameExistsForOtherId(String categoryName, Integer categoryId) {
        return categoryRepository.existsByCategoryNameAndCategoryIdNot(categoryName, categoryId);
    }

    public CategoryResponse createCategory(CategoryDto categoryDto) {
        Category category = new Category();
        category.setCategoryName(categoryDto.getCategoryName());
        Date now = new Date();
        category.setCreatedAt(now);
        category.setUpdatedAt(now);
        return mapToCategoryResponse(categoryRepository.save(category));
    }

    public CategoryResponse updateCategory(Integer categoryId, CategoryDto categoryDto) {
        Optional<Category> optionalCategory = categoryRepository.findById(categoryId);
        if (optionalCategory.isEmpty()) {
            return null;
        }
        Category category = optionalCategory.get();
        category.setCategoryName(categoryDto.getCategoryName());
        category.setUpdatedAt(new Date());
        return mapToCategoryResponse(categoryRepository.save(category));
    }

    public boolean deleteCategory(Integer categoryId) {
        if (!categoryRepository.existsById(categoryId)) {
            return false;
        }
        categoryRepository.deleteById(categoryId);
        return true;
    }

    public Category toggleCategoryStatus(Integer id) {
        Optional<Category> optionalCategory = categoryRepository.findById(id);
        if (optionalCategory.isEmpty()) {
            throw new RuntimeException("Category not found with ID: " + id);
        }
        Category category = optionalCategory.get();
        if (category.getIsActive() && hasActiveSubCategories(id)) {
            throw new IllegalStateException("Không thể ngừng hoạt động danh mục này vì đang có danh mục con hoạt động!");
        }
        category.setIsActive(!category.getIsActive());
        category.setUpdatedAt(new Date());
        return categoryRepository.save(category);
    }

    private boolean hasActiveSubCategories(Integer categoryId) {
        return subCategoryRepository.countByCategoryCategoryIdAndIsActiveTrue(categoryId) > 0;
    }

    private CategoryResponse mapToCategoryResponse(Category category) {
        return new CategoryResponse(
                category.getCategoryId(),
                category.getCategoryName(),
                category.getIsActive(),
                category.getCreatedAt(),
                category.getUpdatedAt()
        );
    }
}

File: frontend/src/services/admin/CategoryService.js
import BaseService from './BaseService';

const API_URL = '/categories';

const CategoryService = {
  async getAllCategories() {
    try {
      const response = await BaseService.get(API_URL);
      return response;
    } catch (error) {
      console.error('Error fetching categories:', error);
      throw error;
    }
  },

  async getCategoryById(id) {
    try {
      const response = await BaseService.get(`${API_URL}/${id}`);
      return response;
    } catch (error) {
      console.error(`Error fetching category with ID ${id}:`, error);
      throw error;
    }
  },

  async createCategory(category) {
    try {
      const response = await BaseService.post(API_URL, category);
      return response;
    } catch (error) {
      console.error('Error creating category:', error);
      throw error;
    }
  },

  async updateCategory(id, category) {
    try {
      const response = await BaseService.put(`${API_URL}/${id}`, category);
      return response;
    } catch (error) {
      console.error(`Error updating category with ID ${id}:`, error);
      throw error;
    }
  },

  async deleteCategory(id) {
    try {
      const response = await BaseService.delete(`${API_URL}/${id}`);
      return response;
    } catch (error) {
      console.error(`Error deleting category with ID ${id}:`, error);
      throw error;
    }
  },

  async toggleCategoryStatus(id) {
    try {
      const response = await BaseService.put(`${API_URL}/${id}/toggle-status`);
      return response;
    } catch (error) {
      console.error(`Error toggling category status with ID ${id}:`, error);
      throw error;
    }
  }
};

export default CategoryService;